### PR TITLE
Update dependencies

### LINKED ISSUE #16: Update `winston`
`molindo-node-logger` currently depends on https://github.com/molindo/molindo-node-logger/blob/ab128078a29041884e50dad7a705162ddfa094d3/package.json#L48

This dependency uses an outdated version of `colors` which prints a warning in webpack 5:

```
[0] warn  - ./node_modules/colors/lib/colors.js
[0] Critical dependency: the request of a dependency is an expression
[0] 
[0] Import trace for requested module:
[0] ./node_modules/colors/lib/colors.js
[0] ./node_modules/colors/safe.js
[0] ./node_modules/winston/lib/winston/config.js
[0] ./node_modules/winston/lib/winston.js
[0] ./node_modules/molindo-node-logger/lib/Logger.js
[0] ./node_modules/molindo-node-logger/lib/index.js
```

We should update the `winston` dependency to get rid of this.

### LINKED ISSUE #17: Use `DEBUG` as the default level for development
Seems more useful based on current usage than `INFO`.

Note that in production we default to printing all levels.

### LINKED ISSUE #19: Remove global listeners
Currently, we register two global listeners:

https://github.com/molindo/molindo-node-logger/blob/03ff4c5a1ad4d46ce779fb47a0def03c8fe709d5/src/Logger.js#L81-L87

This should become the responsibility of apps (if necessary at all). In an environment where loggers are created on-demand, this could potentially leak.

I just saw this in a Next.js dev server:

```
[0] ERROR: (node:96112) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 unhandledRejection listeners added to [process]. Use emitter.setMaxListeners() to increase limit
```

I think this is caused by re-evaluating the logger module during hot reloading when changes are made locally. I don't think this should affect production usage though.

The safest option is definitely to remove this behavior from this library.

### PR DESCRIPTION
**Pull request by symn**
So this was a big one.
Some breaking changes in `winston` and `express-winston` made it a bit harder to update.

Some key things:
- Update winston and express-winston. Closes #16
- There was this internal bug in express-winston, where it would call `winston.log` with the wrong signature: https://github.com/bithavoc/express-winston/issues/209#issuecomment-530129495 - our Logger wrapper relied on this call, so after `express-winston` corrected it, our `createMiddleWare` broke.
- Our interal tests relied on `process.stdout.write` where as now we need to assert on `console._stdout.write`
- Passing `meta` data changed. It took me a bit o find out what is going on, but in the end it was just passing `winston.log(level,message,{meta})` instead of `winston.log(level,message,meta)`
- `winston` removed the `formatter` option and the default `format.simple()` format broke our tests.
- `express-winston` removed the default format on which we asserted so i had to re-add something similar without using lodash's `template` function.
- Supported Node.js version changed to >= 20 (Breaking Change)
- Closes #17 
- Closes #19 

Nevertheless all tests pass and i think we are good to go with this update!

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,9 +1,16 @@
 # Changelog
 
+## 1.3.0
+
+Update devDependencies and dependencies. Update to newest `winston` and
+`express-winston` and don't introduce breaking changes in our logger.
```
Comment (amannn) [inline CHANGELOG.md]: Requiring Node.js 20 is a breaking change, this should be v2. But great that we're still API-compatible!

Comment (amannn) [reply CHANGELOG.md]: As a side note, previous versions in the changelog have referenced the PR, maybe continue doing that?

File: README.md [hunk 2]
```diff
@@ -4,24 +4,24 @@ A Node.js logger that integrates well with the Molindo infrastructure.
...
 
 ## Usage
 
 ```js
 import Logger from 'molindo-node-logger';
 
-const logger = new Logger({service: 'pizza-shop'});
+const logger = new Logger({ service: 'pizza-shop' });
 logger.trace('Making a salami pizza …');
 logger.debug('Adding salami …');
```
Comment (amannn) [inline README.md:22]: Should we at least consider a veggie option to move with the times 😄

File: README.md [hunk 3]
```diff
@@ -4,24 +4,24 @@ A Node.js logger that integrates well with the Molindo infrastructure.
...
+- Offers an express integration that logs HTTP requests, including the `operationName` in the case of GraphQL requests. Confidential headers like `cookie` or `authorization` are masked.
+- Uses the levels and level values from [logback](https://logback.qos.ch/): `ERROR` (40000), `WARN` (30000), `INFO` (20000), `DEBUG` (10000), `TRACE` (5000).
 
 ## Usage
 
 ```js
 import Logger from 'molindo-node-logger';
 
-const logger = new Logger({service: 'pizza-shop'});
+const logger = new Logger({ service: 'pizza-shop' });
```
Comment (amannn) [inline README.md:20]: I've added https://github.com/molindo/eslint-config-molindo/issues/99 to discuss applying our prettier settings to places like Markdown too (not scope of this PR).

File: package.json [hunk 4]
```diff
@@ -12,42 +12,39 @@
     "logging",
     "winston"
   ],
+  "exports": {
+    ".": {
+      "import": "./lib/index.mjs",
```
Comment (amannn) [inline package.json]: Should we move built files to `dist`? I think that's more standard than `lib`.

File: package.json [hunk 5]
```diff
@@ -12,42 +12,39 @@
...
-    "eslint-plugin-css-modules": "2.7.5",
-    "eslint-plugin-import": "2.7.0",
-    "eslint-plugin-import-order-autofix": "0.8.3",
-    "eslint-plugin-jsx-a11y": "6.0.2",
-    "eslint-plugin-prettier": "2.2.0",
-    "eslint-plugin-react": "7.3.0",
-    "jest": "21.2.1",
-    "prettier": "1.7.0",
-    "supertest": "3.0.0"
+    "bunchee": "^5.1.2",
```
Comment (amannn) [inline package.json:32]: 👍

File: package.json [hunk 6]
```diff
@@ -12,42 +12,39 @@
...
+    ".": {
+      "import": "./lib/index.mjs",
+      "default": "./lib/index.js"
+    }
+  },
   "scripts": {
+    "build": "yarn run lint && yarn run test -- --run && bunchee",
     "lint": "eslint src",
-    "test": "jest",
+    "test": "vitest",
```
Comment (amannn) [inline package.json:24]: 👍

File: src/Logger.js [hunk 7]
```diff
@@ -62,81 +63,109 @@ export default class Logger {
...
-          level: this.level
+          stderrLevels: this.stderrLevels
         })
       ],
+      format: this.format(),
+      level: this.level,
       levels: this.winstonLevels,
-      colors: this.colors
+      handleExceptions: false,
+      exitOnError: this.exitOnError
```
Comment (amannn) [inline src/Logger.js]: Is this the right way to do this? I found [`rejectionHandlers`](https://github.com/winstonjs/winston?tab=readme-ov-file#rejections) in the docs. Is `handleExceptions: false` the mechanism to instruct the default transport to be used for unhandled promise rejections?

File: src/Logger.js [hunk 8]
```diff
@@ -62,81 +63,109 @@ export default class Logger {
...
     });
-
-    // winston doesn't allow to set the error level for runtime
-    // exceptions, therefore this should be handled outside of winston.
-    // https://github.com/winstonjs/winston/blob/bd06c4091e9e248e6fed6278e0699e5cfdc7de24/lib/winston/exception-handler.js#L94-L97
-    process.on('uncaughtException', this.onUnhandledException);
-
-    // Unhandled promise rejections should be treated as errors.
-    process.on('unhandledRejection', this.onUnhandledException);
```
Comment (amannn) [inline src/Logger.js:87]: Great that this is gone! Can you add "resolves #19" to the PR description to close https://github.com/molindo/molindo-node-logger/issues/19 when this is merged?

File: src/Logger.js [hunk 9]
```diff
@@ -62,81 +63,109 @@ export default class Logger {
...
         // additional meta data like a logger `name`.
         if (typeof message === 'object') {
           meta = {...message};
           delete meta.message;
           message = message.message;
         }
-
-        this.winston.log(level, message, meta);
+        // Pass the meta object as actual `meta` object
+        this.winston.log(level, message, {meta});
```
Comment (amannn) [inline src/Logger.js:91]: This would be a chance to get rid of `meta`, as [we want to move away from that anyway](https://bitbucket.org/setlist/setlist-widget-festival/pull-requests/23/sfm-2934-forward-request-id#comment-431441666).

But only if you think it's a good time to adapt this, I don't want to scope creep here too much.

File: src/Logger.js [hunk 10]
```diff
@@ -62,81 +63,109 @@ export default class Logger {
...
       };
     });
   }
 
-  onUnhandledException = e => {
+  exitOnError = (e) => {
```
Comment (amannn) [inline src/Logger.js]: Do we need to log here? As far as I understand the docs, this is only intended to negotiate the return value based on an error (which is static in our case anyway).

File: src/Logger.js [hunk 11]
```diff
@@ -62,81 +63,109 @@ export default class Logger {
...
+        requestId
+      };
+
+      info[Symbol.for('message')] = JSON.stringify(log);
+
+      return info;
+    })();
+  }
+
+  formatSimple() {
```
Comment (amannn) [inline src/Logger.js]:
```suggestion
  formatInline() {
```

? "simple" is always such a broad term 🙂

File: src/Logger.js [hunk 12]
```diff
@@ -62,81 +63,109 @@ export default class Logger {
...
+
+      return info;
+    })();
+  }
+
+  formatColorize() {
+    return winston.format.colorize({colors: this.colors});
   }
 
   destroy() {
```
Comment (amannn) [inline src/Logger.js]: Can we remove `destroy()` now that the global handlers are gone? The winston instance should be cleaned up by GC.

File: src/__tests__/Logger.test.js [hunk 13]
```diff
@@ -66,7 +69,8 @@ describe('Logger', () => {
     const logger = new Logger({service: 'pizza-shop', isProduction: true});
     logger.info('pizza is ready!');
 
-    const output = JSON.parse(process.stdout.write.mock.calls[0][0]);
+    const output = JSON.parse(console._stdout.write.mock.calls[0][0]);
+    console.log(output);
```
Comment (amannn) [inline src/__tests__/Logger.test.js]: This is not necessary, right?

File: package.json [hunk 14]
```diff
@@ -12,42 +12,39 @@
...
     "winston"
   ],
+  "exports": {
+    ".": {
+      "import": "./lib/index.mjs",
+      "default": "./lib/index.js"
+    }
+  },
   "scripts": {
+    "build": "yarn run lint && yarn run test -- --run && bunchee",
```
Comment (amannn) [inline package.json]: Lint and test already run as part of `prepublish`, is this necessary?

File: package.json [hunk 15]
```diff
@@ -12,42 +12,39 @@
...
     "lint": "eslint src",
-    "test": "jest",
+    "test": "vitest",
     "test:watch": "npm run test -- --watch --onlyChanged",
-    "build": "babel src -d lib --ignore __tests__",
     "prepublish": "npm run lint && npm run test && rimraf ./lib && npm run build"
```
Comment (amannn) [inline package.json]: Might want to use `yarn run test --run` here.

File: package.json [hunk 16]
```diff
@@ -12,42 +12,39 @@
...
   },
   "dependencies": {
-    "body-parser": "^1.18.2",
-    "express": "^4.16.2",
-    "express-winston": "^2.4.0",
-    "winston": "^2.4.0"
+    "body-parser": "^1.20.2",
+    "express": "^4.19.2",
+    "express-winston": "^4.2.0",
+    "winston": "^3.13.0",
```
Comment (amannn) [inline package.json:43]: Can you reference "Closes #16" in the PR description to auto-close https://github.com/molindo/molindo-node-logger/issues/16 on merge?

File: src/Logger.js [hunk 17]
```diff
@@ -62,81 +63,109 @@ export default class Logger {
       return acc;
     }, {});
 
-    this.winston = new winston.Logger({
+    this.winston = winston.createLogger({
```
Comment (amannn) [inline src/Logger.js:66]: In L52, can we adjust the default level for development? See https://github.com/molindo/molindo-node-logger/issues/17.

**Review by amannn — COMMENTED** [review]
Nice to finally brush off the dust from this package! Did you already have a chance to try out a pre-release in an app?

File: src/Logger.js (hunk 7 above)
Comment (symn) [reply src/Logger.js]: yes i think i was wrong about that.  Property `handleExceptions` is not available at this position, i needs to be in a `transports`: eg.

`new winston.transports.Console({
          stderrLevels: this.stderrLevels,
          handleExceptions: false
        })`
just like we had it before.
I though with `exitOnError` we log the stack trace before exiting?

File: CHANGELOG.md [hunk 18]
```diff
@@ -1,9 +1,20 @@
 # Changelog
 
+## 2.0.0
+
+Update devDependencies and dependencies. Update to newest `winston` and
+`express-winston`. Adapt to breaking changes. ([#20](https://github.com/molindo/molindo-node-logger/pull/20))
+
+### Breaking Changes
+
+- Supported Node.js version changed to >= 20.
```
Comment (amannn) [inline CHANGELOG.md:10]: The removed `destroy` method is a breaking change too.

File: src/Logger.js (hunk 9 above)
Comment (amannn) [reply src/Logger.js:91]: I'll leave this up to you if you want to address this!

File: .eslintrc.js [hunk 19]
```diff
@@ -0,0 +1,12 @@
+module.exports = {
+  extends: ['molindo/javascript'],
+  env: {
+    browser: true,
```
Comment (amannn) [inline .eslintrc.js]: This lib actually doesn't work in the browser.

**Review by amannn — APPROVED** [review]
